### PR TITLE
systemd-logind: allow using BootLoaderEntries DBUS property

### PR DESCRIPTION
**Pull request by fishilico**
`systemd-logind` exposes several properties reload to the bootloader. One of them is `BootLoaderEntries` [1], which scans the disks using util-linux's blkid in order to find the ESP (EFI System Partition) [2][3].

This triggers the following logs in `audit.log` (where `/dev/sda1` is the ESP, mounted on `/boot`):

    type=AVC msg=audit(1577692922.834:310): avc:  denied  { getattr }
    for  pid=690 comm="systemd-logind" name="/" dev="sda1" ino=1
    scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:dosfs_t tclass=filesystem permissive=1

    type=AVC msg=audit(1577692922.841:311): avc:  denied  { search } for
    pid=690 comm="systemd-logind" name="/" dev="sda1" ino=1
    scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:dosfs_t tclass=dir permissive=1

    type=AVC msg=audit(1577692922.841:312): avc:  denied  { getattr }
    for  pid=690 comm="systemd-logind" path="/boot" dev="sda1" ino=1
    scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:dosfs_t tclass=dir permissive=1

    type=AVC msg=audit(1577692922.841:313): avc:  denied  { read } for
    pid=690 comm="systemd-logind" name="sda1" dev="devtmpfs" ino=2496
    scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:fixed_disk_device_t tclass=blk_file
    permissive=1

    type=AVC msg=audit(1577692922.841:313): avc:  denied  { open } for
    pid=690 comm="systemd-logind" path="/dev/sda1" dev="devtmpfs"
    ino=2496 scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:fixed_disk_device_t tclass=blk_file
    permissive=1

    type=AVC msg=audit(1577692922.844:314): avc:  denied  { getattr }
    for  pid=690 comm="systemd-logind" path="/dev/sda1" dev="devtmpfs"
    ino=2496 scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:fixed_disk_device_t tclass=blk_file
    permissive=1

    type=AVC msg=audit(1577692922.844:315): avc:  denied  { ioctl } for
    pid=690 comm="systemd-logind" path="/dev/sda1" dev="devtmpfs"
    ino=2496 ioctlcmd=0x1272 scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:fixed_disk_device_t tclass=blk_file
    permissive=1

    type=AVC msg=audit(1577692922.851:316): avc:  denied  { read } for
    pid=690 comm="systemd-logind" name="loader.conf" dev="sda1" ino=4
    scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:dosfs_t tclass=file permissive=1

    type=AVC msg=audit(1577692922.851:316): avc:  denied  { open } for
    pid=690 comm="systemd-logind" path="/boot/loader/loader.conf"
    dev="sda1" ino=4 scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:dosfs_t tclass=file permissive=1

    type=AVC msg=audit(1577692922.851:317): avc:  denied  { getattr }
    for  pid=690 comm="systemd-logind" path="/boot/loader/loader.conf"
    dev="sda1" ino=4 scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:dosfs_t tclass=file permissive=1

    type=AVC msg=audit(1577692922.851:318): avc:  denied  { ioctl } for
    pid=690 comm="systemd-logind" path="/boot/loader/loader.conf"
    dev="sda1" ino=4 ioctlcmd=0x5401
    scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:dosfs_t tclass=file permissive=1

    type=AVC msg=audit(1577692922.851:319): avc:  denied  { read } for
    pid=690 comm="systemd-logind" name="entries" dev="sda1" ino=5
    scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:dosfs_t tclass=dir permissive=1

    type=AVC msg=audit(1577692922.851:319): avc:  denied  { open } for
    pid=690 comm="systemd-logind" path="/boot/loader/entries" dev="sda1"
    ino=5 scontext=system_u:system_r:systemd_logind_t
    tcontext=system_u:object_r:dosfs_t tclass=dir permissive=1

[1] https://github.com/systemd/systemd/blob/v244/src/login/logind-dbus.c#L3315
[2] https://github.com/systemd/systemd/blob/v244/src/login/logind-dbus.c#L3118
[3] https://github.com/systemd/systemd/blob/v244/src/shared/bootspec.c#L835

### CODE REVIEW

File: policy/modules/system/systemd.te [hunk 1]
```diff
@@ -487,6 +490,7 @@ selinux_get_enforce_mode(systemd_logind_t)
 
 storage_getattr_removable_dev(systemd_logind_t)
 storage_getattr_scsi_generic_dev(systemd_logind_t)
+storage_raw_read_fixed_disk(systemd_logind_t)
```
Comment (pebenito) [inline policy/modules/system/systemd.te]: These rules should be conditional, especially because of this.

Comment (fishilico) [reply policy/modules/system/systemd.te]: I updated my commit in order to try introducing a tunable named `systemd_logind_get_bootloader`, but this fails to build:
```
/usr/bin/checkmodule -m tmp/systemd.tmp -o tmp/systemd.mod
policy/modules/system/systemd.te:566:
ERROR 'syntax error' at token 'typeattribute' on line 54942:
	typeattribute systemd_logind_t fixed_disk_raw_read;
#line 566
/usr/bin/checkmodule:  error(s) encountered while parsing configuration
/usr/bin/checkmodule:  loading policy configuration from tmp/systemd.tmp
```

As far as I understand, there is a `neverallow` rule which makes attribute `fixed_disk_raw_read` mandatory in order to read fixed disks (and this is nice), but associating types to attributes cannot be conditional/tunable. How would you like to handle this?

**Comment by doverride** [comment]
Nicolas Iooss <notifications@github.com> writes:

> @fishilico commented on this pull request.
>
> ------------------------------------------------------------------------------------------------------------------------------------------------
>
> In policy/modules/system/systemd.te:
>
>> @@ -487,6 +490,7 @@ selinux_get_enforce_mode(systemd_logind_t)
>  
>  storage_getattr_removable_dev(systemd_logind_t)
>  storage_getattr_scsi_generic_dev(systemd_logind_t)
> +storage_raw_read_fixed_disk(systemd_logind_t)
>
> I updated my commit in order to try introducing a tunable named systemd_logind_get_bootloader, but this fails to build:
>
> /usr/bin/checkmodule -m tmp/systemd.tmp -o tmp/systemd.mod
> policy/modules/system/systemd.te:566:
> ERROR 'syntax error' at token 'typeattribute' on line 54942:
> 	typeattribute systemd_logind_t fixed_disk_raw_read;
> #line 566
> /usr/bin/checkmodule:  error(s) encountered while parsing configuration
> /usr/bin/checkmodule:  loading policy configuration from tmp/systemd.tmp
>
> As far as I understand, there is a neverallow rule which makes attribute fixed_disk_raw_read mandatory in order to read fixed disks (and this
> is nice), but associating types to attributes cannot be conditional/tunable. How would you like to handle this?

You arent allowed to "typeattributeset" in booleans.

That effectively means that storage_raw_read_fixed_disk() should be
renamed to storage_raw_read_fixed_disk_uncond()

Then you would create a storage_raw_read_fixed_disk() without the
"typeattribute $1 fixed_disk_raw_read;"

and create a seperate interface for the "typeattribute $1
fixed_disk_raw_read;"

You would call the storage_raw_read_fixed_disk() from the boolean and
the interface with "typeattribute $1 fixed_disk_raw_read;"
unconditionally

To see existing examples: grep -r uncond policy/modules/*/*.if

>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

-- 
Key fingerprint = 5F4D 3CDB D3F8 3652 FBD8 02D5 3B6C 5F1D 2C7B 6B02
https://sks-keyservers.net/pks/lookup?op=get&search=0x3B6C5F1D2C7B6B02
Dominick Grift

**Comment by fishilico** [comment]
The refpolicy currently uses interfaces with both `_cond` and `_uncond` suffixes. Here is a commit that:
* does not rename `storage_raw_read_fixed_disk`,
* follows the idea of associating a domain with attribute `fixed_disk_raw_read` (that does not grant anything but validating a `neverallow` rule),
* adds a single interface to add the attribute and allow reading fixed disks in a tunable block, as I find the result easier to read.

**Pull request merged**